Services/ProtocolServer/GeminiDownload.cpp

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2020, The SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <LibGemini/GeminiResponse.h>
+#include <LibGemini/GeminiJob.h>
+#include <ProtocolServer/GeminiDownload.h>
+
+GeminiDownload::GeminiDownload(PSClientConnection& client, NonnullRefPtr<Gemini::GeminiJob>&& job)
+    : Download(client)
+    , m_job(job)
+{
+    m_job->on_finish = [this](bool success) {
+        if (auto* response = m_job->response()) {
+            set_payload(response->payload());
+            if (!response->meta().is_empty()) {
+                HashMap<String, String, CaseInsensitiveStringTraits> headers;
+                headers.set("meta", response->meta());
+                set_response_headers(headers);
+            }
+        }
+
+        // if we didn't know the total size, pretend that the download finished successfully
+        // and set the total size to the downloaded size
```

Comment:
Does it actually pretend it finished successfully? To me this reads like it just reports a 100% loading progress, not any kind of success.

Reply:
the proceeding `did_finish` implies success

Reply:
This comment brought to you by CTRL+C and CTRL+V, in association with: https://github.com/SerenityOS/serenity/blob/3c3c1b40131b41454c59c4f5c8f6eaf9094556a1/Services/ProtocolServer/HttpsDownload.cpp#L41-L44